File: ruby/word-count/word_count.rb
class Phrase
  WORD_PATTERN = /\b[\w']+\b/
  def initialize(phrase)
    @phrase = phrase
  end

  def word_count
    words(@phrase).reduce(Hash.new(0)) {|hash, word| hash[word] += 1; hash}
  end

  def words(phrase)
    phrase.downcase.scan(WORD_PATTERN)
  end

end

File: elixir/word-count/word_count.exs
defmodule Words do
  @doc """
  Count the number of words in the sentence.

  Words are compared case-insensitively.
  """
  @spec count(String.t()) :: map
  def count(sentence) do
    sentence
    |> String.downcase()
    |> String.split(~r([^-\p{L}\d}]+)u, trim: true)
    |> Enum.reduce(%{}, fn key, map -> Map.update(map, key, 1, &(&1 + 1)) end)
  end
end
